### PR TITLE
Add additional resource checks to upgrade testing

### PR DESCRIPTION
**Pull request by markusewalker**
## Problem

When we perform fresh install provisioning in the Go framework, we perform additional checks after the cluster is provisioned. These are general validation checks we perform manually alongside actual cluster provisioning. If and when QA performs upgrades on their clusters, we perform the same validation checks as well.

In our automation, we do not have those same additional resource checks when upgrading. This PR adds those checks.
 
## Solution

Added additional resource checks to `kubernetes_test.go`, so that after a single cluster upgrade, we are covering the same checks that we have in the fresh install use-cases.
 
### Manual Testing

Tested RKE1 and K3S and validated that the resource checks are working as expected, after an upgrade.

### Automated Testing

Jenkins jobs will be provided to the reviewers offline.

### CODE REVIEW

File: tests/v2/validation/upgrade/README.md [hunk 1]
```diff
@@ -7,12 +7,12 @@ In your config file, set the following, this will run each test in parallel both
 ```yaml
```
Comment (caliskanugur) [inline tests/v2/validation/upgrade/README.md:7]: I don't see any difference for the items in the array, the previous one makes it a bit more obvious IMO because the cluster name comes first, what do you think?

File: tests/v2/validation/upgrade/kubernetes_test.go [hunk 2]
```diff
@@ -106,4 +110,23 @@ func (u *UpgradeKubernetesTestSuite) testUpgradeSingleCluster(clusterName, versi
...
+	assert.NotEmpty(u.T(), podResults)
+	assert.Empty(u.T(), podErrors)
+
+	if psact == string(RancherPrivileged) || psact == string(RancherRestricted) {
+		err = psadeploy.CheckPSACT(client, clusterName)
+		require.NoError(u.T(), err)
+
+		_, err = psadeploy.CreateNginxDeployment(client, clusterMeta.ID, psact)
+		require.NoError(u.T(), err)
+	}
```
Comment (caliskanugur) [inline tests/v2/validation/upgrade/kubernetes_test.go:131]: A question here, this is applicable for the current provisioning tests as well: Does making PSACT checks before the pods' check cover more? Cause PSACT creates another deployment if enabled

**Review by caliskanugur — COMMENTED** [review]
Looks good, got a few questions, PR needs a freeform that includes all supported cluster types

File: tests/v2/validation/upgrade/README.md (hunk 1 above)
Comment (markusewalker) [reply tests/v2/validation/upgrade/README.md:7]: I can revert it back, I think I originally modified it because there was some indentation issues, but I'll revert it back

File: tests/v2/validation/upgrade/kubernetes_test.go (hunk 2 above)
Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go:131]: I'm going to add a check actually. If deployment name of `nginx` already exists, I'll change the deployment name. Just so that we don't even have a possibility of a name conflict.

Comment (caliskanugur) [reply tests/v2/validation/upgrade/kubernetes_test.go:131]: Yes that's what I mean, in upgrade HA, provisioning tests and this test are chained, I think it might make it fail

Comment (caliskanugur) [inline tests/v2/validation/upgrade/kubernetes_test.go:131]:
```suggestion
	if psact == string(RancherPrivileged) || psact == string(RancherRestricted) {
		err = psadeploy.CheckPSACT(client, clusterName)
		require.NoError(u.T(), err)

		_, err = psadeploy.CreateNginxDeployment(client, clusterMeta.ID, psact)
		require.NoError(u.T(), err)
	}
	
	podResults, podErrors := pods.StatusPods(client, clusterMeta.ID)
	assert.NotEmpty(u.T(), podResults)
	assert.Empty(u.T(), podErrors)
```

This is the second part of what I was asking, wouldn't the pod status check cover nginx deployment as well?

Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go:131]: Yeah, I have a new commit I will push in a moment that should address 😄

Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go:131]: It can, we had that check how it is because we want to check the pods in the system workload page as part of one of the team's general validation checks. Since the PSACT check already checks the state, technically it does that.

It's not a big deal to me either way to put it after, but I'd have to do that for all the other files too which would make the PR a bit bigger

Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go:131]: Resolved. Check if `nginx` already exists and if it does, create a new one with the prefix `upgrade-`.

File: tests/framework/extensions/psact/createdeployment.go [hunk 3]
```diff
@@ -38,6 +38,10 @@ func CreateNginxDeployment(client *rancher.Client, clusterID string, psact strin
 		return nil, err
 	}
 
+	if _, err := steveclient.SteveType(workloads.DeploymentSteveType).ByID(deploymentTemplate.Namespace + "/" + deploymentTemplate.Name); err == nil {
+		deploymentTemplate.Name = deploymentTemplate.Name + "-2"
```
Comment (slickwarren) [inline tests/framework/extensions/psact/createdeployment.go:42]: why is this necessary?

File: tests/framework/extensions/clusters/clusters.go [hunk 4]
```diff
@@ -869,7 +869,7 @@ func isClusterInaccessible(messages []string) (isInaccessible bool) {
 }
 
 func logClusterInfoWithChanges(clusterID, clusterInfo string, summary summary.Summary) string {
-	newClusterInfo := fmt.Sprintf("ClusterID: %v, Message: %v, Error: %v, State: %v, Transiationing: %v", clusterID, summary.Message, summary.Error, summary.State, summary.Transitioning)
+	newClusterInfo := fmt.Sprintf("ClusterID: %v, Message: %v, Error: %v, State: %v, Transitioning: %v", clusterID, summary.Message, summary.Error, summary.State, summary.Transitioning)
```
Comment (slickwarren) [inline tests/framework/extensions/clusters/clusters.go:872]: thanks for fixing this, just saw it pop up in a separate test run :)

File: tests/v2/validation/upgrade/kubernetes_test.go [hunk 5]
```diff
@@ -106,4 +110,23 @@ func (u *UpgradeKubernetesTestSuite) testUpgradeSingleCluster(clusterName, versi
 		u.T().Logf("[%v]: Validating updated cluster's nodepools kubernetes versions", clusterName)
 		validateNodepoolVersions(u.T(), client, updatedCluster, version, !isCheckingCurrentCluster)
 	}
+
+	err = nodestat.IsNodeReady(client, clusterMeta.ID)
```
Comment (slickwarren) [inline tests/v2/validation/upgrade/kubernetes_test.go:114]: imo we should add something like 
```
err = wait.WatchWait(result, checkFunc)
assert.NoError(t, err)
assert.Equal(t, clusterName, clusterResp.ObjectMeta.Name)
```
to also check that the cluster is in a ready state. In my testing, I found that all nodes ready != cluster ready.

File: tests/v2/validation/upgrade/kubernetes_test.go [hunk 6]
```diff
@@ -106,4 +110,23 @@ func (u *UpgradeKubernetesTestSuite) testUpgradeSingleCluster(clusterName, versi
...
+
+	podResults, podErrors := pods.StatusPods(client, clusterMeta.ID)
+	assert.NotEmpty(u.T(), podResults)
+	assert.Empty(u.T(), podErrors)
+
+	if psact == string(RancherPrivileged) || psact == string(RancherRestricted) {
+		err = psadeploy.CheckPSACT(client, clusterName)
+		require.NoError(u.T(), err)
+
+		_, err = psadeploy.CreateNginxDeployment(client, clusterMeta.ID, psact)
```
Comment (slickwarren) [inline tests/v2/validation/upgrade/kubernetes_test.go:129]: in the create function, can you add the deployment name to the log? Currently reads as `"Deployment successfully created; this is expected for rancher-privileged!"`

**Review by slickwarren — COMMENTED** [review]
are any of these mutually exclusive from what we would check on a freshly created cluster? If not, it would make more sense to me to have a general (extension?) function to call that does all post-healthy cluster checks. Some other nits below;

File: tests/framework/extensions/psact/createdeployment.go (hunk 3 above)
Comment (markusewalker) [reply tests/framework/extensions/psact/createdeployment.go:42]: In the scenario where you run the upgrade test immediately after you do the fresh install provisioning automation, there will be a name conflict and the test fails. This gets around that issue.

File: tests/v2/validation/upgrade/kubernetes_test.go (hunk 5 above)
Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go:114]: I can, but that leads to a point made in an earlier comment that it'd better to make that change for all of the other files. I don't necessarily have an issue with that, but it will make the scope of this PR larger.

File: tests/v2/validation/upgrade/kubernetes_test.go (hunk 6 above)
Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go:129]: Resolved.

File: tests/framework/extensions/psact/createdeployment.go (hunk 3 above)
Comment (slickwarren) [reply tests/framework/extensions/psact/createdeployment.go:42]: can you add a comment in the code for this special case?

File: tests/v2/validation/upgrade/kubernetes_test.go (hunk 5 above)
Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go:114]: Further thinking of this, is this even necessary though? I mean, we have checks in the automation already to see if a cluster is active before we check for nodes status. Would that address your concern or am I misunderstanding?

File: tests/v2/validation/upgrade/kubernetes_test.go (hunk 2 above)
Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go]: Resolved. Made the change to this file and the other provisioning files that need the change.

File: tests/v2/validation/upgrade/kubernetes_test.go (hunk 5 above)
Comment (slickwarren) [reply tests/v2/validation/upgrade/kubernetes_test.go:114]: you're right, this is done earlier in this test.

**Review by slickwarren — APPROVED** [review]
lgtm

File: tests/framework/extensions/psact/createdeployment.go (hunk 3 above)
Comment (caliskanugur) [reply tests/framework/extensions/psact/createdeployment.go:43]: Would it fail in this case: provision once, upgrade twice, or more? If yes, would making this prefix random make it re-runnable in any case? If not, please ignore this comment

**Review by caliskanugur — APPROVED** [review]
Just a question and a nit, looks good to me

File: tests/framework/extensions/psact/createdeployment.go (hunk 3 above)
Comment (caliskanugur) [reply tests/framework/extensions/psact/createdeployment.go:43]: Yes, we've got a name generator package in the framework: https://github.com/rancher/rancher/blob/release/v2.7/tests/framework/pkg/namegenerator/namegenerator.go#L17-L25

Comment (markusewalker) [reply tests/framework/extensions/psact/createdeployment.go:43]: Forgot about that, I'll make a change in a new commit, thanks Ugur!

Comment (markusewalker) [reply tests/framework/extensions/psact/createdeployment.go]: Addressed.

File: tests/v2/validation/upgrade/kubernetes_test.go [hunk 7]
```diff
@@ -106,4 +110,23 @@ func (u *UpgradeKubernetesTestSuite) testUpgradeSingleCluster(clusterName, versi
 		u.T().Logf("[%v]: Validating updated cluster's nodepools kubernetes versions", clusterName)
 		validateNodepoolVersions(u.T(), client, updatedCluster, version, !isCheckingCurrentCluster)
 	}
+
+	err = nodestat.IsNodeReady(client, clusterMeta.ID)
+	require.NoError(u.T(), err)
```
Comment (caliskanugur) [inline tests/v2/validation/upgrade/kubernetes_test.go:115]: So I wrote this comment to a resolved comment, I agree with you, is this check necessary since we already check the cluster's active state before it gets out of the polls?

Comment (markusewalker) [reply tests/v2/validation/upgrade/kubernetes_test.go:115]: I would say so, because this checks the node's status. I've seen times while manually testing where just because the cluster is active, it doesn't guarantee the nodes are active as well.

**Review by caliskanugur — APPROVED** [review]
Looks good to me

**Review by slickwarren — DISMISSED** [review]
lgtm

**Review by igomez06 — APPROVED** [review]

**Review by igomez06 — APPROVED** [review]
